lib/checks/checks.go

```diff
@@ -310,163 +308,155 @@ type Server struct {
 		return nil
 	}
 
 	var errors []error
+
 	// check each server against its profile
-	for _, server := range r.servers {
-		requirements := r.requirements[server.Server.Role]
-		validateCtx, cancel := context.WithTimeout(ctx, defaults.AgentValidationTimeout)
-		defer cancel()
-		failed, err := r.remote.Validate(validateCtx, server.AdvertiseIP, r.manifest, server.Server.Role)
-		if err != nil {
-			log.Warnf("Failed to validate remote node: %v.", trace.DebugReport(err))
-			errors = append(errors,
-				trace.BadParameter("failed to validate remote node %v", server))
-		}
-		if len(failed) != 0 {
-			errors = append(errors, trace.BadParameter("%v failed checks:\n%v",
-				server, FormatFailedChecks(failed)))
-		}
+	for _, server := range r.Servers {
+		errors = append(errors, r.RunSingleNodeChecks(ctx, server)...)
+	}
 
-		err = checkServerProfile(server, requirements)
-		if err != nil {
-			errors = append(errors, err)
-		}
+	// run checks that take all servers into account
+	errors = append(errors, r.RunMultiNodeChecks(ctx, r.Servers)...)
 
-		dockerConfig := r.manifest.SystemDocker()
-		if r.TestDockerDevice {
-			err = checkDockerDevice(server, dockerConfig)
-			if err != nil {
-				errors = append(errors, err)
-			}
-		}
+	return trace.NewAggregate(errors...)
+}
 
-		err = checkSystemPackages(server, dockerConfig)
-		if err != nil {
-			errors = append(errors, err)
-		}
+// RunSingleNodeChecks executes checks for the provided individual server.
+func (r *checker) RunSingleNodeChecks(ctx context.Context, server Server) (errors []error) {
+	if ifTestsDisabled() {
+		log.Infof("Skipping single-node checks due to %q set.",
+			constants.PreflightChecksOffEnvVar)
+		return nil
+	}
+
+	requirements := r.Reqs[server.Server.Role]
+	validateCtx, cancel := context.WithTimeout(ctx, defaults.AgentValidationTimeout)
+	defer cancel()
+	failed, err := r.Remote.Validate(validateCtx, server.AdvertiseIP, ValidateConfig{
+		Manifest: r.Manifest,
+		Profile:  server.Server.Role,
+		Docker:   requirements.Docker,
+	})
+	if err != nil {
+		log.Warnf("Failed to validate remote node: %v.", trace.DebugReport(err))
+		errors = append(errors,
+			trace.BadParameter("failed to validate remote node %v", server))
+	}
+	if len(failed) != 0 {
+		errors = append(errors, trace.BadParameter("%v failed checks:\n%v",
+			server, FormatFailedChecks(failed)))
+	}
 
-		err = r.checkTempDir(ctx, server)
+	err = checkServerProfile(server, requirements)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	dockerConfig := r.Manifest.SystemDocker()
+	if r.TestDockerDevice {
+		err = checkDockerDevice(server, dockerConfig)
 		if err != nil {
 			errors = append(errors, err)
 		}
 	}
 
-	// run checks that take all servers into account
-	err := checkSameOS(r.servers)
+	err = checkSystemPackages(server, dockerConfig)
 	if err != nil {
 		errors = append(errors, err)
 	}
 
-	err = checkTime(time.Now().UTC(), r.servers)
+	err = r.checkTempDir(ctx, server)
```

Comment:
Btw, temporary directory check seems to be broken as temporary directory is not reported by agents so this basically tests as if the temporary directory == working directory of the process.
I'll create a ticket for this.